0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+Many target languages will have dedicated types designed to hold any Unicode scalar value,
+such as the `char` primitive type in Rust or the `rune` alias for `int32` in Go.
+
+The main complexity in implementing this change is minimizing the migration pain for
+code bases that use `{:extern}` definitions to pass string values in and out of the Dafny runtime.
+This generally requires converting between the internal representation of Dafny strings
+and the native implementation in the target language,
+e.g. between `Dafny.ISequence<char>` and `System.String` in C#.
+Even there, the various Dafny runtime libraries provide helper methods that can be adjusted 
+to work with the new semantics of Unicode characters.
+In the C# runtime, for example, the `Sequence.FromString(string)` method converts a native string
+to a equivalent `Dafny.ISequence<char>` copy.
+A parallel method named something similar to `Sequence.UnicodeFromString(string)` could be added
+that would return a `Dafny.ISequence<uint32>` copy instead.
+Migrating an existing codebase should reduce to a simple find-and-replace operation.
+
+Each runtime should also make it possible to wrap an idiomatic representation of a Unicode string
+as the runtime representation of the Dafny `seq<char>` type, without having to make an additional copy. 
+At the time of writing this, only the Java runtime supports this feature,
+via a `StringDafnySequence` class that wraps a `java.lang.String` value and implements the `ISequence<Character>` interface.
+Another similar parallel `UnicodeStringDafnySequence` adaptor can be provided that implements `ISequence<Integer>` instead.
+
+The one wrinkle with this feature is that methods such as `select(i)`, which implement Dafny expressions such as `s[i]`,
+currently take constant time on an `StringDafnySequence`, since a `java.lang.String` also stores its contents as UTF-16
+code units. However, if `UnicodeStringDafnySequence.select(i)` will need to produce the i'th Unicode scalar value instead,
+this will require decoding all characters up to that index, which will take linear time in general instead.
+The solution will be to lazily decode strings into their UTF-32 encoded equivalent (i.e. sequence of scalar values)
+on the first operation that requires it. This mirrors the optimization of evaluating concatenation of sequences
+lazily in the current C# and Java runtimes, and means that in most code that manipulates individual characters,
+the amortized runtime of indexing operations will still be constant.
+
+Note that this optimization will also help offset the fact that materializing and manipulating individual `char` values
+will in general require twice as much memory after this change.
+The compiler can leverage this support to initially store string literal values as UTF-8 bytes,
+and operations such as concatenation or equality comparisons can be implemented
+without having to decode these bytes into character values.
```

Comment:
The proposal does not address distinguishing `seq<char>` and `string` at the language level.  Do we want to keep `string` an alias of `seq<char>`?   Will that not make our lives harder at the compiler level? (I'm worried about e.g. polymorphic functions over `seq<T>`.

Reply:
That being said, if we change `seq<T>` from a type to a trait, it would be nice to make `string` implement `seq<char>`

Reply:
Split this off as a separate discussion: https://github.com/dafny-lang/dafny/issues/2582

Reply:
This is one point I didn't burden the proposal with, but I'll say two things:

* https://github.com/dafny-lang/dafny/issues/2582 makes me think that we don't really need to distinguish `string` from `seq<T>` at runtime in general.
* I'd say that being able to define `seq<T>` in Dafny ala https://github.com/robin-aws/dafny/blob/dafny-runtime-in-dafny/Source/DafnyRuntime/DafnyRuntimeDafny/src/dafnyRuntime.dfy#L382 will provide much of the same benefit as literally making `seq<T>` a trait directly.

Reply:
> I'd say that being able to define seq<T> in Dafny … will provide much of the same benefit as literally making seq<T> a trait directly.

Yep, 100% this